### DISCUSSION TITLE
2FA redirection logic

### DISCUSSION THREAD

**Discussion by fayeinmay** [authelia/authelia#7703] [Q&A] [answered]
Hi,

I enabled 2FA for a specific subdomain. If a user in the 'guest' group (as seen below) now visits my authelia domain (auth.mydomain.de), they get redirected to /2fa/one-time-password. If 2FA is configured, but not enabled for any subdomains, the users get redirected to /authenticated. I looked at the source code and it seems like this is intended behavior, but I wasn't entirely sure as it seems odd to me to request a user to enable 2FA, even though they have no access to any 2FA protected page anyway. So is this intended, or a misconfiguration of some sort?

![image](https://github.com/user-attachments/assets/61002167-7321-4670-be5f-e78b7b724320)
results in
![image](https://github.com/user-attachments/assets/fa7cf9a9-e596-44cc-8bb8-ee561a0cf7c8)

**Comment by james-d-elliott** [comment]
Authelia will redirect with proxies that support it in the following scenarios from memory:

1. The user is not authenticated but needs to be.
2. The user is not sufficiently authenticated to the resource and they can potentially be authenticated sufficiently based on ACLs.

Proxies like Traefik and Caddy, and I believe HAProxy fully support this redirection behavior as they allow quite literally forwarding the responses to the client directly in the case of a non 2xx. 

How nginx handles this is entirely different. They are very limited in their ability to forward the response to the client, in fact you can't forward the response at all and have to have nginx write its own response. In addition the only codes it allows are `200`, `401`, and `403` making it extremely limiting and frankly flawed. The only way to handle this specific proxy is to handle the 401/403 error codes and either display content (i.e. error) or perform a redirect entirely controlled on the proxy side. I believe this is what you're running into.

**Reply by fayeinmay** [reply]
Hmm, I think the way I asked my question was misleading. I'm totally fine with the redirection of my reverse proxy. I just wondered if it is intentional to display this 2FA message on the Login Portal page in case that any subdomain requires 2FA, even though the user does not have any access to any 2FA pages so it wouldn't even make sense to them adding 2FA to their account.

Im referring to https://github.com/authelia/authelia/blob/6461c2f10d68355ca71a2b78a69ba32d30b2e009/web/src/views/LoginPortal/LoginPortal.tsx#L133

**Reply by james-d-elliott** [reply]
The available methods value is the list of methods available for the user to use for 2FA, not the methods they have configured. For example the instance may not be configured with Duo or may have TOTP or WebAuthn disabled.

**Reply by james-d-elliott** [reply]
There's also one condition where the 2FA registration is always shown if there are methods available to the user (to add or otherwise), when they authenticate directly at the portal without a redirection.

**Reply by fayeinmay** [reply]
Huh, but why is the list different in case I change two_factor to one_factor in this configuration? The user is in the group 'guest' and the redirection changes from /2fa/one-time-password to /authenticated when changing two_factor to one_factor for the group 'fox".

![358867162-61002167-7321-4670-be5f-e78b7b724320](https://github.com/user-attachments/assets/cd12ff86-1d1f-4778-a0ec-054b36f26311)

**Reply by fayeinmay** [reply]
> There's also one condition where the 2FA registration is always shown if there are methods available to the user (to add or otherwise), when they authenticate directly at the portal without a redirection.

This is what I was referring to! So it is intended after all.

**Reply by james-d-elliott** [reply, marked as answer]
Yes absolutely. So if they're visiting authelia without context of where they want to go, we don't redirect them unless there is a default redirection URL and they're fully authenticated. The reason for this is if we did then there would be a bad UX if they ever wanted to manage authelia settings.

**Reply by fayeinmay** [reply]
True. And if no policy rule is using two_factor, you just put them to /authenticated as they would then not need to engage with 2FA settings. That makes sense.
